File: HeartbeatCalculator.swift
// HeartbeatCalculator.swift
// INFINITY OS · 库存心跳系统 v2.0
//
// 核心公式：
//   a = r × T  （心跳单元 = 日销率 × 周期天数）
//   水位 = 总库存 ÷ a

import Foundation

struct HeartbeatCalculator {

    // 水位状态
    enum WaterStatus {
        case safe       // ≥ 3a
        case normal     // 2a ~ 3a
        case warning    // 1a ~ 2a
        case danger     // < 1a

        var color: String {
            switch self {
            case .safe, .normal: return "#4ADE80"
            case .warning:       return "#F5A623"
            case .danger:        return "#FF5F5F"
            }
        }

        var badgeText: String {
            switch self {
            case .safe:    return "✅ 安全水位"
            case .normal:  return "🟡 接近预警"
            case .warning: return "⚠️ 补货预警"
            case .danger:  return "🔴 危险缺货"
            }
        }
    }

    // 计算结果
    struct Result {
        let dailySalesRate: Double      // r：日销率
        let cycleT: Int                 // T：周期天数
        let heartbeatUnit: Double       // a = r × T
        let shelfBottles: Int           // 货架库存（瓶）
        let warehouseBottles: Int       // 仓库库存（瓶，已换算）
        let totalBottles: Int           // 合计
        let waterLevel: Double          // 水位倍数
        let status: WaterStatus         // 状态
        let replenishNeeded: Int        // 建议补货（瓶）
        let replenishCases: Int         // 建议补货（箱）
        let advice: String              // 补货建议文字
        let formulaDetail: String       // 公式展示
    }

    // 门店等级 → 周期T
    static func cycleT(for grade: String) -> Int {
        switch grade {
        case "S+", "S", "A": return 7
        case "B": return 14
        case "C": return 21
        case "D": return 30
        default:  return 14
        }
    }

    // 核心计算
    static func calculate(
        dailySalesRate: Double,
        grade: String,
        shelfQty: Int,
        warehouseQty: Int,
        bottlesPerCase: Int
    ) -> Result {

        let T = cycleT(for: grade)
        let a = dailySalesRate * Double(T)           // 心跳单元
        let warehouseBottles = warehouseQty * bottlesPerCase
        let total = shelfQty + warehouseBottles
        let waterLevel = a > 0 ? Double(total) / a : 0.0

        // 判断状态
        let status: WaterStatus
        let replenishNeeded: Int
        let replenishCases: Int
        let advice: String

        if waterLevel >= 3.0 {
            status = .safe
            replenishNeeded = 0
            replenishCases = 0
            let daysLeft = dailySalesRate > 0 ? Double(shelfQty) / dailySalesRate : 0
            advice = String(format: "库存充足，货架%d瓶约%.1f天消耗，正常跟进。", shelfQty, daysLeft)
        } else if waterLevel >= 2.0 {
            status = .normal
            let daysToWarning = dailySalesRate > 0
                ? (Double(total) - a * 2.0) / dailySalesRate : 0
            replenishNeeded = 0
            replenishCases = 0
            advice = String(format: "库存正常，约%.0f天后将触及预警线，注意观察。", daysToWarning)
        } else if waterLevel >= 1.0 {
            status = .warning
            replenishNeeded = Int(ceil(a * 3.0 - Double(total)))
            replenishCases = Int(ceil(Double(replenishNeeded) / Double(bottlesPerCase)))
            advice = String(format: "⚠️ 已触发补货预警！建议立即补货 %d瓶（%d箱），补至3a满仓水位。",
                replenishNeeded, replenishCases)
        } else {
            status = .danger
            replenishNeeded = Int(ceil(a * 3.0 - Double(total)))
            replenishCases = Int(ceil(Double(replenishNeeded) / Double(bottlesPerCase)))
            advice = String(format: "🔴 严重缺货！货架领地已失守。紧急补货 %d瓶（%d箱）。",
                replenishNeeded, replenishCases)
        }

        let formulaDetail = String(format:
            "r = %.0f瓶/天\nT = %d天（%@级周期）\na = r×T = %.0f×%d = %.0f瓶\n货架%d瓶 + 仓库%d瓶 = 合计%d瓶\n水位 = %d ÷ %.0f = %.2fa\n3a满仓线 = %.0f瓶  |  2a预警线 = %.0f瓶",
            dailySalesRate, T, grade,
            dailySalesRate, T, a,
            shelfQty, warehouseBottles, total,
            total, a, waterLevel,
            a * 3, a * 2)

        return Result(
            dailySalesRate: dailySalesRate,
            cycleT: T,
            heartbeatUnit: a,
            shelfBottles: shelfQty,
            warehouseBottles: warehouseBottles,
            totalBottles: total,
            waterLevel: waterLevel,
            status: status,
            replenishNeeded: replenishNeeded,
            replenishCases: replenishCases,
            advice: advice,
            formulaDetail: formulaDetail
        )
    }
}

File: HeartbeatCalculator.java
package com.xinhe.sfa.inventory;

/**
 * 心跳公式计算器
 * INFINITY OS · 库存心跳系统 v2.0
 *
 * 核心公式：
 *   a = r × T  （心跳单元 = 日销率 × 周期天数）
 *   水位 = 总库存 ÷ a
 *
 * 三层水位：
 *   ≥ 3a = 满仓（安全）
 *   2a ~ 3a = 正常
 *   1a ~ 2a = 预警，触发补货
 *   < 1a = 危险，货架领地失守
 */
public class HeartbeatCalculator {

    // 门店等级对应的周期T（天）
    public static int getTByGrade(String grade) {
        switch (grade) {
            case "S+": return 7;
            case "S":  return 7;
            case "A":  return 7;
            case "B":  return 14;
            case "C":  return 21;
            case "D":  return 30;
            default:   return 14;
        }
    }

    // 门店等级对应的拜访频次描述
    public static String getVisitFrequency(String grade) {
        switch (grade) {
            case "S+": return "3次/周";
            case "S":  return "2次/周";
            case "A":  return "1次/周";
            case "B":  return "每两周";
            case "C":  return "每三周";
            case "D":  return "每月";
            default:   return "未知";
        }
    }

    /**
     * 计算结果数据类
     */
    public static class Result {
        public float dailySalesRate;   // r：日销率（瓶/天）
        public int cycleT;             // T：周期天数
        public float heartbeatUnit;    // a：心跳单元（瓶）
        public int shelfBottles;       // 货架库存（瓶）
        public int warehouseBottles;   // 仓库库存（瓶，已换算）
        public int totalBottles;       // 总库存（瓶）
        public float waterLevel;       // 水位（a的倍数）
        public WaterStatus status;     // 水位状态
        public int replenishNeeded;    // 建议补货量（瓶）
        public int replenishCases;     // 建议补货量（箱）
        public String advice;          // 补货建议文字
        public String formulaDetail;   // 公式计算过程（用于展示）

        public enum WaterStatus {
            SAFE,       // ≥ 3a 安全
            NORMAL,     // 2a ~ 3a 正常
            WARNING,    // 1a ~ 2a 预警
            DANGER      // < 1a 危险
        }
    }

    /**
     * 核心计算方法
     *
     * @param dailySalesRate  日销率 r（瓶/天）
     * @param grade           门店等级（S+/S/A/B/C/D）
     * @param shelfQty        货架库存（瓶）
     * @param warehouseQty    仓库库存（箱）
     * @param bottlesPerCase  每箱瓶数
     * @return 计算结果
     */
    public static Result calculate(
            float dailySalesRate,
            String grade,
            int shelfQty,
            int warehouseQty,
            int bottlesPerCase) {

        Result result = new Result();

        // 基础参数
        result.dailySalesRate = dailySalesRate;
        result.cycleT = getTByGrade(grade);
        result.shelfBottles = shelfQty;
        result.warehouseBottles = warehouseQty * bottlesPerCase;
        result.totalBottles = result.shelfBottles + result.warehouseBottles;

        // 心跳单元 a = r × T
        result.heartbeatUnit = dailySalesRate * result.cycleT;

        // 水位 = 总库存 ÷ a
        if (result.heartbeatUnit > 0) {
            result.waterLevel = result.totalBottles / result.heartbeatUnit;
        } else {
            result.waterLevel = 0;
        }

        // 判断水位状态
        if (result.waterLevel >= 3.0f) {
            result.status = Result.WaterStatus.SAFE;
            result.replenishNeeded = 0;
            result.replenishCases = 0;
            result.advice = String.format(
                "库存充足，货架%.0f瓶约%.1f天消耗，正常跟进。",
                (float) shelfQty,
                shelfQty / Math.max(dailySalesRate, 0.1f));
        } else if (result.waterLevel >= 2.0f) {
            result.status = Result.WaterStatus.NORMAL;
            int daysToWarning = (int) ((result.totalBottles - result.heartbeatUnit * 2) / Math.max(dailySalesRate, 0.1f));
            result.replenishNeeded = 0;
            result.replenishCases = 0;
            result.advice = String.format("库存正常，约%d天后将触及预警线，注意观察。", daysToWarning);
        } else if (result.waterLevel >= 1.0f) {
            result.status = Result.WaterStatus.WARNING;
            result.replenishNeeded = (int) Math.ceil(result.heartbeatUnit * 3 - result.totalBottles);
            result.replenishCases = (int) Math.ceil((float) result.replenishNeeded / bottlesPerCase);
            result.advice = String.format(
                "⚠️ 已触发补货预警！建议立即补货 %d瓶（%d箱），补至3a满仓水位。",
                result.replenishNeeded, result.replenishCases);
        } else {
            result.status = Result.WaterStatus.DANGER;
            result.replenishNeeded = (int) Math.ceil(result.heartbeatUnit * 3 - result.totalBottles);
            result.replenishCases = (int) Math.ceil((float) result.replenishNeeded / bottlesPerCase);
            result.advice = String.format(
                "🔴 严重缺货！货架领地已失守。紧急补货 %d瓶（%d箱）。",
                result.replenishNeeded, result.replenishCases);
        }

        // 公式展示文字
        result.formulaDetail = String.format(
            "r = %.0f瓶/天\nT = %d天（%s级周期）\na = r×T = %.0f×%d = %.0f瓶\n货架%d瓶 + 仓库%d瓶 = 合计%d瓶\n水位 = %d ÷ %.0f = %.2fa\n3a满仓线 = %.0f瓶  |  2a预警线 = %.0f瓶",
            dailySalesRate,
            result.cycleT, grade,
            dailySalesRate, result.cycleT, result.heartbeatUnit,
            shelfQty, result.warehouseBottles, result.totalBottles,
            result.totalBottles, result.heartbeatUnit, result.waterLevel,
            result.heartbeatUnit * 3, result.heartbeatUnit * 2);

        return result;
    }
}
